models/camelot/camelot_pair_fee_rates.sql

```diff
@@ -0,0 +1,163 @@
+            token1,
+            {{ v3_default_fee }} / {{ v3_fee_precision }} as token1_fee_rate
+        from {{ source("camelot_v3_arbitrum", "AlgebraFactory_evt_Pool") }}
+    ),
+    v3_directional_fee_rate_updates as (
+        select
+            date_trunc('minute', evt_block_time) as minute,
+            pair,
+            version,
+            token0,
+            feezto / {{ v3_fee_precision }} as token0_fee_rate,
+            token1,
+            feeotz / {{ v3_fee_precision }} as token1_fee_rate
+        from {{ source("camelot_v3_arbitrum", "AlgebraPool_evt_Fee") }} as fee_updates
+        join
+            v3_pairs_with_initial_fee_rates as pairs
+            on fee_updates.contract_address = pairs.pair
+    ),
+    pairs as (
+        select *
+        from v2_pairs_with_initial_fee_rates
+        union all
+        select *
+        from v3_pairs_with_initial_fee_rates
+    ),
+    fee_rate_updates as (
+        select
+            minute,
+            pair,
+            version,
+            token0,
+            avg(token0_fee_rate) as token0_fee_rate,  -- Handle edge case where pair fees gets changed multiple times per minute
+            token1,
+            avg(token1_fee_rate) as token1_fee_rate  -- Handle edge case where pair fees gets changed multiple times per minute
+        from
+            (
+                select *
+                from pairs
+                union all
+                select *
+                from v2_directional_fee_rate_updates
+                union all
+                select *
+                from v3_directional_fee_rate_updates
+            )
+        group by minute, pair, version, token0, token1
+    ),
+    camelot_pair_trades_by_minute as (
+        select distinct
+            date_trunc('minute', block_time) as minute, project_contract_address as pair
+        from dex.trades
+        where
+            blockchain = '{{blockchain}}' and project = 'camelot'
+            {% if not is_incremental() %}
+                and block_time >= timestamp '{{project_start_date}}'
+            {% endif %}
+            {% if is_incremental() %}
+                and block_time >= date_trunc('day', now() - interval '7' day)
+            {% endif %}
+    ),
+    -- Prepare data structure (1 row for every minute where pair trades happened
+    -- and/or fee rates got updated)
+    pairs_by_minute as (
+        select distinct minute, pair
+        from
+            (
+                select minute, pair
+                from camelot_pair_trades_by_minute
+                union all
+                select minute, pair
+                from fee_rate_updates
+            )
+    ),
+    -- Add version, token0, token1 columns
+    pairs_by_minute_with_metadata as (
+        select pairs_by_minute.minute, pairs_by_minute.pair, version, token0, token1
+        from pairs_by_minute
+        left join pairs on pairs_by_minute.pair = pairs.pair
```

Comment:
i was hoping we could avoid the left join back, but i don't think it's possible. only other option is to do left join in dex CTE where there is less data (pre union step) to grab necessary columns, then union fully in the union distinct above with all columns you need

Reply:
not possible imo, since we also need the metadata for non-trades (pair deployments + fee updates), which also produce rows in the final result set